File: 02.py
#!/usr/bin/env python3
from functools import reduce, partial
from sys import stdin
from math import prod

look = { "forward": lambda x: (x,0), "up": lambda x: (0,-x), "down": lambda x: (0,x) }
ls   = [(l[0], int(l[1])) for l in (l.strip().split() for l in stdin)]
xs   = [sum(col) for col in zip(*(look.get(k)(x) for k,x in ls))]
look = { "forward": lambda x,a,b,c: (a,    b+x, c+a*x),
         "up":      lambda x,a,b,c: (-x+a, b,   c),
         "down":    lambda x,a,b,c: (x+a,  b,   c) }
ys   = (partial(look.get(k), x) for k,x in ls)
ys   = reduce(lambda a, f: f(*a), ys, (0,0,0))

print (f"Part 1: { prod(xs) }, Part 2: { prod(ys[1:]) }")

File: 02.scala
#!/usr/bin/env scala
import scala.io.Source.stdin

@main def aoc(): Unit = 
  val ns = stdin.getLines().map(_.split(' ')).map(x=>(x(0).head, x(1).toInt)).toList
  val f  = (k:Char,v:Int) => k match { case 'f'=>(v,0) case 'u'=>(0,-v) case 'd'=>(0,v) }
  val p1 = ns.map(f.tupled).reduce((a,b)=>(a._1+b._1, a._2+b._2)).toList
  val p2 = ns.foldLeft((0,0,0))((a,b)=>b._1 match {
    case 'f'=>(a._1,       a._2+b._2, a._3+a._1*b._2)
    case 'u'=>(-b._2+a._1, a._2,      a._3)
    case 'd'=>(b._2+a._1,  a._2,      a._3) }).toList

  println(s"Part 1: ${ p1.product }, Part 2: ${ p2.tail.product }")